DataDog nil pointer dereference

## Summary 

User reported the following panic in logs:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x5a9ee2]

goroutine 227 [running]:
net/url.(*URL).Query(0x0, 0x2)
	/usr/local/go/src/net/url/url.go:1110 +0x22
github.com/argoproj/argo-rollouts/metricproviders/datadog.(*Provider).Run(0xc0006fc1e0, 0xc000da64e0, 0xc0008e4600, 0x1b, 0xc000cbad00, 0x2, 0x0, 0x0, 0x0, 0xc000cbad02, ...)
	/go/src/github.com/argoproj/argo-rollouts/metricproviders/datadog/datadog.go:86 +0x169
github.com/argoproj/argo-rollouts/analysis.(*Controller).runMeasurements.func1(0xc000e9e4a4, 0xc000da64e0, 0x2e07888, 0x0, 0x0, 0xc000e9e4b0, 0xc00044cf30, 0xc000ea6700, 0xc0008e4600, 0x1b, ...)
	/go/src/github.com/argoproj/argo-rollouts/analysis/analysis.go:280 +0xe35
created by github.com/argoproj/argo-rollouts/analysis.(*Controller).runMeasurements
	/go/src/github.com/argoproj/argo-rollouts/analysis/analysis.go:251 +0x278
```

## Diagnostics

What version of Argo Rollouts are you running? unsure

Here was affected code. It is likely caused by misconfiguration, but we should handle it.


```golang
func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
	startTime := metav1.Now()

	// Measurement to pass back
	measurement := v1alpha1.Measurement{
		StartedAt: &startTime,
	}

	endpoint := "https://api.datadoghq.com/api/v1/query"
	if p.config.Address != "" {
		endpoint = p.config.Address + "/api/v1/query"
	}

	url, _ := url.Parse(endpoint)

	now := unixNow()
	var interval int64 = 300
	if metric.Provider.Datadog.Interval != "" {
		expDuration, err := metric.Provider.Datadog.Interval.Duration()
		if err != nil {
			return metricutil.MarkMeasurementError(measurement, err)
		}
		// Convert to seconds as DataDog expects unix timestamp
		interval = int64(expDuration.Seconds())
	}

	q := url.Query() // panic
```



---
<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

--- a/USERS.md
+++ b/USERS.md
@@ -6,6 +6,7 @@ Organizations below are **officially** using Argo Rollouts. Please send a PR wit
 1. [Ambassador Labs](https://www.getambassador.io)
 1. [Ant Group](https://www.antgroup.com/)
 1. [Bucketplace](https://www.bucketplace.co.kr/)
+1. [Codefresh](https://codefresh.io/)
 1. [Databricks](https://github.com/databricks)
 1. [Devtron Labs](https://github.com/devtron-labs/devtron)
 1. [Intuit](https://www.intuit.com/)

--- a/metricproviders/datadog/datadog.go
+++ b/metricproviders/datadog/datadog.go
@@ -70,7 +70,10 @@ func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alph
 		endpoint = p.config.Address + "/api/v1/query"
 	}
 
-	url, _ := url.Parse(endpoint)
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
 
 	now := unixNow()
 	var interval int64 = 300

--- a/metricproviders/datadog/datadog_test.go
+++ b/metricproviders/datadog/datadog_test.go
@@ -27,6 +27,7 @@ func TestRunSuite(t *testing.T) {
 
 	// Test Cases
 	var tests = []struct {
+		serverURL               string
 		webServerStatus         int
 		webServerResponse       string
 		metric                  v1alpha1.Metric
@@ -172,62 +173,86 @@ func TestRunSuite(t *testing.T) {
 			expectedErrorMessage:    "Could not parse JSON body: json: cannot unmarshal string into Go struct field datadogResponse.Series of type []struct { Pointlist [][]float64 \"json:\\\"pointlist\\\"\" }",
 			useEnvVarForKeys:        false,
 		},
+
+		// Error if server address is faulty
+		{
+			serverURL: "://wrong.schema",
+			metric: v1alpha1.Metric{
+				Name:             "foo",
+				SuccessCondition: "result < 0.001",
+				FailureCondition: "result >= 0.001",
+				Provider: v1alpha1.MetricProvider{
+					Datadog: &v1alpha1.DatadogMetric{
+						Query: "avg:kubernetes.cpu.user.total{*}",
+					},
+				},
+			},
+			expectedPhase:        v1alpha1.AnalysisPhaseError,
+			expectedErrorMessage: "parse \"://wrong.schema/api/v1/query\": missing protocol scheme",
+			useEnvVarForKeys:     false,
+		},
 	}
 
 	// Run
 
 	for _, test := range tests {
-		// Server setup with response
-		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		serverURL := test.serverURL
 
-			//Check query variables
-			actualQuery := req.URL.Query().Get("query")
-			actualFrom := req.URL.Query().Get("from")
-			actualTo := req.URL.Query().Get("to")
+		if serverURL == "" {
+			// Server setup with response
+			server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
 
-			if actualQuery != "avg:kubernetes.cpu.user.total{*}" {
-				t.Errorf("\nquery expected avg:kubernetes.cpu.user.total{*} but got %s", actualQuery)
-			}
+				//Check query variables
+				actualQuery := req.URL.Query().Get("query")
+				actualFrom := req.URL.Query().Get("from")
+				actualTo := req.URL.Query().Get("to")
 
-			if from, err := strconv.ParseInt(actualFrom, 10, 64); err == nil && from != unixNow()-test.expectedIntervalSeconds {
-				t.Errorf("\nfrom %d expected be equal to %d", from, unixNow()-test.expectedIntervalSeconds)
-			} else if err != nil {
-				t.Errorf("\nfailed to parse from: %v", err)
-			}
+				if actualQuery != "avg:kubernetes.cpu.user.total{*}" {
+					t.Errorf("\nquery expected avg:kubernetes.cpu.user.total{*} but got %s", actualQuery)
+				}
 
-			if to, err := strconv.ParseInt(actualTo, 10, 64); err == nil && to != unixNow() {
-				t.Errorf("\nto %d was expected be equal to %d", to, unixNow())
-			} else if err != nil {
-				t.Errorf("\nfailed to parse to: %v", err)
-			}
+				if from, err := strconv.ParseInt(actualFrom, 10, 64); err == nil && from != unixNow()-test.expectedIntervalSeconds {
+					t.Errorf("\nfrom %d expected be equal to %d", from, unixNow()-test.expectedIntervalSeconds)
+				} else if err != nil {
+					t.Errorf("\nfailed to parse from: %v", err)
+				}
 
-			//Check headers
-			if req.Header.Get("Content-Type") != "application/json" {
-				t.Errorf("\nContent-Type header expected to be application/json but got %s", req.Header.Get("Content-Type"))
-			}
-			if req.Header.Get("DD-API-KEY") != expectedApiKey {
-				t.Errorf("\nDD-API-KEY header expected %s but got %s", expectedApiKey, req.Header.Get("DD-API-KEY"))
-			}
-			if req.Header.Get("DD-APPLICATION-KEY") != expectedAppKey {
-				t.Errorf("\nDD-APPLICATION-KEY header expected %s but got %s", expectedAppKey, req.Header.Get("DD-APPLICATION-KEY"))
-			}
+				if to, err := strconv.ParseInt(actualTo, 10, 64); err == nil && to != unixNow() {
+					t.Errorf("\nto %d was expected be equal to %d", to, unixNow())
+				} else if err != nil {
+					t.Errorf("\nfailed to parse to: %v", err)
+				}
 
-			// Return mock response
-			if test.webServerStatus < 200 || test.webServerStatus >= 300 {
-				http.Error(rw, test.webServerResponse, test.webServerStatus)
-			} else {
-				rw.Header().Set("Content-Type", "application/json")
-				io.WriteString(rw, test.webServerResponse)
-			}
-		}))
-		defer server.Close()
+				//Check headers
+				if req.Header.Get("Content-Type") != "application/json" {
+					t.Errorf("\nContent-Type header expected to be application/json but got %s", req.Header.Get("Content-Type"))
+				}
+				if req.Header.Get("DD-API-KEY") != expectedApiKey {
+					t.Errorf("\nDD-API-KEY header expected %s but got %s", expectedApiKey, req.Header.Get("DD-API-KEY"))
+				}
+				if req.Header.Get("DD-APPLICATION-KEY") != expectedAppKey {
+					t.Errorf("\nDD-APPLICATION-KEY header expected %s but got %s", expectedAppKey, req.Header.Get("DD-APPLICATION-KEY"))
+				}
+
+				// Return mock response
+				if test.webServerStatus < 200 || test.webServerStatus >= 300 {
+					http.Error(rw, test.webServerResponse, test.webServerStatus)
+				} else {
+					rw.Header().Set("Content-Type", "application/json")
+					io.WriteString(rw, test.webServerResponse)
+				}
+			}))
+			defer server.Close()
+
+			serverURL = server.URL
+		}
 
 		tokenSecret := &corev1.Secret{
 			ObjectMeta: metav1.ObjectMeta{
 				Name: DatadogTokensSecretName,
 			},
 			Data: map[string][]byte{
-				"address": []byte(server.URL),
+				"address": []byte(serverURL),
 				"api-key": []byte(expectedApiKey),
 				"app-key": []byte(expectedAppKey),
 			},
@@ -236,7 +261,7 @@ func TestRunSuite(t *testing.T) {
 		if test.useEnvVarForKeys {
 			os.Setenv("DD_API_KEY", expectedApiKey)
 			os.Setenv("DD_APP_KEY", expectedAppKey)
-			os.Setenv("DD_ADDRESS", server.URL)
+			os.Setenv("DD_ADDRESS", serverURL)
 		} else {
 			os.Unsetenv("DD_API_KEY")
 			os.Unsetenv("DD_APP_KEY")

--- a/metricproviders/datadog/datadog_test.go
+++ b/metricproviders/datadog/datadog_test.go
@@ -176,17 +176,8 @@ func TestRunSuite(t *testing.T) {
 
 		// Error if server address is faulty
 		{
-			serverURL: "://wrong.schema",
-			metric: v1alpha1.Metric{
-				Name:             "foo",
-				SuccessCondition: "result < 0.001",
-				FailureCondition: "result >= 0.001",
-				Provider: v1alpha1.MetricProvider{
-					Datadog: &v1alpha1.DatadogMetric{
-						Query: "avg:kubernetes.cpu.user.total{*}",
-					},
-				},
-			},
+			serverURL:            "://wrong.schema",
+			metric:               v1alpha1.Metric{},
 			expectedPhase:        v1alpha1.AnalysisPhaseError,
 			expectedErrorMessage: "parse \"://wrong.schema/api/v1/query\": missing protocol scheme",
 			useEnvVarForKeys:     false,

